library/core/src/intrinsics/bounds.rs

```diff
@@ -0,0 +1,36 @@
+//! Various traits used to restrict intrinsics to not-completely-wrong types.
+
+/// Types with a built-in dereference operator in runtime MIR,
+/// aka references and raw pointers.
+pub trait BuiltinDeref: Sized {
+    type Pointee: ?Sized;
+}
+
+impl<T: ?Sized> BuiltinDeref for &mut T {
+    type Pointee = T;
+}
+impl<T: ?Sized> BuiltinDeref for &T {
+    type Pointee = T;
+}
+impl<T: ?Sized> BuiltinDeref for *mut T {
+    type Pointee = T;
+}
+impl<T: ?Sized> BuiltinDeref for *const T {
+    type Pointee = T;
+}
+
+pub trait ChangePointee<U: ?Sized>: BuiltinDeref {
```

Comment:
this trait is *purely* type-projection hijinx, by comparison, so I think it's fine for it to be safe.